plutus-tx/src/PlutusTx/Extensions/Secrets.hs

```diff
@@ -0,0 +1,133 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TypeApplications  #-}
+module PlutusTx.Extensions.Secrets(
+  Secret
+  , SecretArgument (..)
+  , mkSecret
+  , secretArg
+  , extractSecret
+  , escape_sha2_256
+  , unsafe_escape_secret
+  ) where
+
+import           Control.Monad
+import           Data.Aeson                   as Aeson (FromJSON (..), ToJSON (..), Value (..))
+import           Data.Aeson.Encoding.Internal (string)
+import           PlutusTx.Prelude             as PlutusTx
+import           Prelude                      as Haskell
+
+-- | A secret value. A value of type `Secret a` can't leak onto
+-- the blockchain in plain-text unless you use an unsafe function.
+-- However, a value of type `Secret a` can end up on the blockchain
+-- via one of the escape hatches like `escape_sha2_256`.
+newtype Secret a = MkSecret a
+
+{- Note [Secret abstraction]
+ - A value of type `Secret a` is guaranteed not to leak (without a compiler warning)
```

Comment:
This is an interesting formatting of multiline comments! You really don't need all these leading `-`s :)

Reply:
Applies throughout.

Reply:
To be clear, our usual formatting is:
```
{- A title maybe
text
more text
-}
```

Reply:
Fixed.